src/Connection/UpgradedStream.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Amp\Http\Client\Connection;
+
+use Amp\ByteStream\InputStream;
+use Amp\Failure;
+use Amp\Http\Client\HttpException;
+use Amp\Http\Client\Internal\ForbidCloning;
+use Amp\Http\Client\Internal\ForbidSerialization;
+use Amp\Promise;
+use Amp\Socket\Socket;
+use Amp\Socket\SocketAddress;
+use Amp\Socket\SocketException;
+use function Amp\call;
+
+final class UpgradedStream implements Socket
+{
+    use ForbidCloning;
+    use ForbidSerialization;
+
+    private $read;
+    private $write;
+    private $localAddress;
+    private $remoteAddress;
+
+    public function __construct(InputStream $read, callable $write, SocketAddress $local, SocketAddress $remote)
```

Comment:
With the `callable` I prefer this class to be internal, I guess.

Reply:
I guess `$write` Should just be an `OutputStream`?

Reply:
For a class in the public namespace, yes it should use `OutputStream` instead. Probably the simplest way to do that is wrap the callable in an internal class implementing `OutputStream`.